File: lab1-333/main.c
// Author: Gao Junran
// Date: 2024-03-26
// Description: This is a program to convert a floating point number to its IEEE 754 representation.
// My program is tested in https://baseconvert.com/ieee-754-floating-point, and turns out to be accurate.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MANTISSA_SIZE 24
#define EXPONENT_SIZE 8
#define MAX_SIZE 200

typedef struct {
    char sign[2];
    char mantissa[MANTISSA_SIZE + 1];
    char exponent[EXPONENT_SIZE + 1];
} F32Repr;


void pad_zero(char *source, char *target, int size, int isLeading) {
    int len = strlen(source); 

    if (len >= size) {
        strcpy(target, source); 
        return;
    }

    int padCount = size - len; 

    if (isLeading) {
        for (int i = 0; i < padCount; i++) {
            target[i] = '0'; 
        }
        strcpy(target + padCount, source); 
    } else {
        strcpy(target, source); 
        for (int i = 0; i < padCount; i++) {
            target[len + i] = '0'; 
        }
        target[size] = '\0'; 
    }
}

void int_to_binary(long long int num, char *binary) {
    int index = 0;
    char temp[MAX_SIZE] = {0};
    while (num > 0) {
        temp[index++] = (num % 2) + '0';
        num /= 2;
    }
    int len = index;
    for (int i = 0; i < len; i++) {
        binary[i] = temp[len - i - 1];
    }
    binary[len] = '\0';
}

void dec_to_binary(long double dec, char *binary) {
    int index = 0;
    if (dec == 0.0) {
        binary[index++] = '0';
    }
    while (dec > 0.0 && index < 32) {
        dec *= 2.0;
        if (dec >= 1.0) {
            binary[index++] = '1';
            dec -= 1.0;
        } else {
            binary[index++] = '0';
        }
    }
    binary[index] = '\0';
}

void calculate_mantissa(char *int_binary, char *dec_binary, char *mantissa) {
    char binary[MAX_SIZE] = {0};
    strcpy(binary, int_binary);
    strcat(binary, dec_binary);
    while (binary[0] == '0' && binary[1] != '\0') {
        memmove(binary, binary + 1, strlen(binary));
    }
    if (binary[0] == '\0') {
        strcpy(binary, "1");
    }
    pad_zero(binary, mantissa, MANTISSA_SIZE, 0);  // pad trailing zeros
    mantissa[MANTISSA_SIZE] = '\0';
}

void calculate_exponent(long double real, char *exponent) {
    int exp_val = 0;
    if (real >= 1.0) {
        while (real >= 2.0) {
            real /= 2.0;
            exp_val++;
        }
    } else {
        while (real < 1.0) {
            real *= 2.0;
            exp_val--;
        }
    }
    exp_val += 127;
    if (exp_val > 255) {
        exp_val = 255;  // ensure exponent does not exceed 8 bits
    }
    char temp[MAX_SIZE];
    int_to_binary(exp_val, temp);
    pad_zero(temp, exponent, EXPONENT_SIZE, 1);  // pad leading zeros
}

int main() {
    printf("Input a real number: "); fflush(stdout);
    long double real;
    scanf("%Lf", &real);
    
    F32Repr repr;
    repr.sign[0] = (real >= 0.0) ? '0' : '1';
    repr.sign[1] = '\0';
    real = fabs(real);
    
    long long int int_part = (long long int) real;
    long double dec_part = real - int_part;
    
    char int_binary[MAX_SIZE] = {0}, dec_binary[MAX_SIZE] = {0};
    int_to_binary(int_part, int_binary);
    dec_to_binary(dec_part, dec_binary);
    
    calculate_mantissa(int_binary, dec_binary, repr.mantissa);
    calculate_exponent(real, repr.exponent);
    
    if (strcmp(repr.exponent, "11111111") == 0) {
        memset(repr.mantissa, '0', MANTISSA_SIZE);
        repr.mantissa[MANTISSA_SIZE] = '\0';
        printf("The IEEE 754 Single Precision is \nSign: %s\nMantissa (24 bit): %s\nExponent: %s\n", repr.sign, repr.mantissa, repr.exponent);
        fflush(stdout);
        printf("Overflow!!!\n"); fflush(stdout);
    } else {
        printf("The IEEE 754 Single Precision is \nSign: %s\nMantissa (24 bit): %s\nExponent: %s\n", repr.sign, repr.mantissa, repr.exponent);
        fflush(stdout);
    }
    return 0;
}

File: lab1-333/src/main.rs
/// Author: Gao Junran
/// Date: 2024-03-26
/// Description: This is a program to convert a floating point number to its IEEE 754 representation.
/// My program is tested in https://baseconvert.com/ieee-754-floating-point, and turns out to be accurate.
use std::io::stdin;
use std::io::stdout;
use std::io::Write;

struct F32Repr {
    sign: String,
    mantissa: String,
    exponent: String,
}

impl F32Repr {
    fn new() -> Self {
        Self {
            sign: String::new(),
            mantissa: String::new(),
            exponent: String::new(),
        }
    }

    fn repr(&self) -> String {
        format!(
            "The IEEE 754 Single Precision is \nSign: {}\nMantissa (24 bit): {}\nExponent: {}",
            self.sign, self.mantissa, self.exponent
        )
    }
}

fn int_to_binary(int: i64) -> String {
    let mut integer = int;
    let mut binary = String::new();
    while integer > 0 {
        let remainder = integer % 2;
        binary.push_str(&remainder.to_string());
        integer /= 2;
    }
    binary.chars().rev().collect::<String>()
}

fn dec_to_binary(dec: f64) -> String {
    let mut decimal = dec;
    let mut binary = String::new();
    if decimal == 0.0 {
        binary.push('0');
    }
    while decimal > 0.0 {
        decimal *= 2.0;
        if decimal >= 1.0 {
            binary.push('1');
            decimal -= 1.0;
        } else {
            binary.push('0');
        }
    }
    binary
}

/// From slide: page 46
fn calculate_mantissa(int_binary: &String, dec_binary: &String) -> String {
    let mut binary = format!("{}{}", int_binary, dec_binary);
    // drop the leading zeros
    while binary.starts_with("0") {
        binary.remove(0);
    }
    // consider: 0.0
    if binary.is_empty() {
        binary.push('1');
    }
    binary = format!("{:0<24}", binary);
    binary[..24].to_string()
}

/// From slide: page 47
fn calculate_exponent(real: f64) -> String {
    let mut real = real;
    let mut exponent = 0;
    if real >= 1.0 {
        // For greater than 1: keep dividing by 2 until it's less than 2
        while real >= 2.0 {
            real /= 2.0;
            exponent += 1;
        }
    } else {
        // For less than 1: keep multiplying by 2 until it's greater than 1
        while real < 1.0 {
            real *= 2.0;
            exponent -= 1;
        }
    }
    exponent += 127; // bias
    if exponent > 255 {
        // overflow
        exponent = 255;
    }
    format!("{:0>8}", int_to_binary(exponent))
}

fn main() {
    println!("Input a real number: ");
    stdout().flush().unwrap();
    let mut real = String::new();
    stdin().read_line(&mut real).unwrap();
    // parse to float first to accept scientific notation
    let mut real = real.trim().parse::<f64>().unwrap();
    let mut repr = F32Repr::new();
    repr.sign = if real >= 0.0 { "0" } else { "1" }.to_string();
    // After the sign is determined, make the number positive
    real = real.abs();
    let (int, dec) = (
        real as i64,  // integer part
        real.fract(), // decimal part
    );
    let int_binary = int_to_binary(int);
    let dec_binary = dec_to_binary(dec);
    repr.mantissa = calculate_mantissa(&int_binary, &dec_binary);
    repr.exponent = calculate_exponent(real);
    if repr.exponent == "11111111" {
        repr.mantissa = "000000000000000000000000".to_string(); // rule by IEEE 754
        println!("{}", repr.repr());
        println!("Overflow!!!")
    } else {
        println!("{}", repr.repr());
    }
}
